### DISCUSSION TITLE
아이템 60. 정확한 답이 필요하다면 float와 double은 피하라

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Effective-Java/Effective-Java#145] [3. 과제]
📝 구성

- [0. TL;DR 🔨](#0-tl-dr---)
- [1. 금융 계산에 부동소수 타입의 사용](#1-------------------)
- [2. BigDecimal을 사용한 해법](#2-bigdecimal--------)
- [3. 정수 타입을 사용한 해법](#3--------------)
- [4. 핵심 요약 📚](#4---------)
- [5. 회고 🧹](#5------)

<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>

---

# 0. TL;DR 🔨

- 소수점의 정확한 연산을 위함이라면 float과 double을 `지양` 하라.
- BigDecimal의 `반올림 모드`는 정확한 연산을 돕는다.
- 성능 향상과 소수점 직접 추적이 가능하다면 int와 long을 사용하라.

---

# 1. 금융 계산에 부동소수 타입의 사용

float과 double은 부동소수를 사용한다. 
금융계산처럼 명확해야 하는 계산에는 지양해야한다.

아래와 같은 코드가 있다.

- 금융 계산에 부동소수 타입을 사용

``` java
    private static void floatingCalculate() {
        double funds = 1.00;
        int itemBought = 0;

        for (double price = 0.10; funds >= price; price += 0.10) {
            funds -= price;
            itemBought++;
        }

        System.out.println(itemBought + "개 구입");
        System.out.println("잔돈(달러) : " + funds);
    }
```

기대한 결과는 사탕 1개 -> 사탕 2개 -> 사탕 3개 -> 사탕 4개 의 구매를 통해 1 + 2 + 3 + 4의 결과로 1달러의 소비를 원했다.

하지만 실제 결과는
![image](https://user-images.githubusercontent.com/53285909/226150127-7714bb74-3d96-4c3d-a8a1-ee0155f71f8d.png)

좀 더 계산이 명확해야한다.

# 2. BigDecimal을 사용한 해법

```java
    private static void bigDecimalCalculate() {
        final BigDecimal TEN_CENTS = new BigDecimal(".10");

        int itemBought = 0;

        BigDecimal funds = new BigDecimal(1.00);

        for (BigDecimal price = TEN_CENTS; funds.compareTo(price) >= 0; price = price.add(TEN_CENTS)) {
            funds = funds.subtract(price);
            itemBought++;
        }

        System.out.println(itemBought + "개 구입");
        System.out.println("잔돈(달러) : " + funds);
    }
```

![image](https://user-images.githubusercontent.com/53285909/226150367-a48d805a-eb91-4908-9b95-ca0c9de2bfcb.png)

부동소수점의 사용과 달리 명확한 계산결과가 반환된다.
하지만 BigDecimal은 두가지 단점이 있다.

1. 불편하다.
2. 느리다.

이를 대안할 수 있는 방법이 있는데...

# 3. 정수 타입을 사용한 해법

위의 식을 다시 생각 해보면 다룰 수 있는 값의 크기가 제한할 수있다. (꼭 소수점을 사용하지 않아도 계산이 된다.)
즉 아래와 같은 코드로 작성이 가능하다.

```java
    private static void integerCalculate() {
        int itemBought = 0;
        int funds = 100;

        for (int price = 10; funds >= price; price += 10) {
            funds -= price;
            itemBought++;
        }

        System.out.println(itemBought + "개 구입");
        System.out.println("잔돈(센트) : " + funds);
    }
```

결과는 아래와 같다.
![image](https://user-images.githubusercontent.com/53285909/226150497-4f5bd81d-7398-4ecf-905d-301843d7541a.png)

# 4. 핵심 요약 📚

- 위의 예제는 각각의 상황에 알맞게 사용 해야한다.
- 정확한 소수점이 필요한 계산이라면 반드시 BigDecimal을 사용해야한다.
- 3번의 예제는 소수점 계산이 필요하지 않은 계산인지를 한번 더 체크하자는 의미에서 작성된 것 같다.

> int는 아홉 자리 십진수
> long 열여덟 자리 십진수
> 열여덟 자리 십진수 이상은 BigDecimal 사용

# 5. 회고 🧹

2023.03.19 일

- 코테 하다보면 소수점 계산에서 생각보다 화를 입기 쉽상이다. 그리고 어쩔때는 double 정도로 해결되는 문제들도 많았지만
- 공간 복잡도, 시간 복잡도에 크게 구애받지 않는다면 BigDecimald을 사용해 해결해 볼 수 있도록 해야겠다.

**Comment by corock** [comment]
잘 봤습니다~! 회고 쪽에 오타가 있네요(`BigDecimald`* -> `BigDecimal`).

이번 주 제 아이템을 정리하다가 `java.text.NumberFormat` 클래스가 있더라구요.
금액 관련 사용 시에 이 클래스를 활용해보는 것도 괜찮을 것 같습니다.
